### PR TITLE
Added .py and notebook to predict sales

### PR DESCRIPTION
**Pull request by JCardenasRdz**
Added code to predict sales based on campaign features  for the wine data set.

Project 18.

### CODE REVIEW

File: Projects/18_Clustering/Predict_Sales.py [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+    # get values
+    cat = df[col].values;
+    # learn categories
+    le.fit(cat)
+    # encode into float and allocate
+    df[col] = le.transform(cat)
+
+# extract data 
+y = df.Total_Sales
+X = df.iloc[:,0:5]
```
Comment (amirziai) [inline Projects/18_Clustering/Predict_Sales.py]: Always add a newlines at the end of files. http://stackoverflow.com/questions/729692/why-should-text-files-end-with-a-newline

File: Projects/18_Clustering/Predict_Sales.py [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+df_total_sales = df_total_sales.groupby(['offer_id']).sum()
+df_total_sales.columns = ['Total_Sales']
+df_total_sales = df_total_sales.reset_index()
+
+# create final DF
+df_offers_and_sales = pd.merge(df_offers,df_total_sales)
+
+# encode categorical predictors
+from sklearn import preprocessing
+le = preprocessing.LabelEncoder()
```
Comment (amirziai) [inline Projects/18_Clustering/Predict_Sales.py]: Using `LabelEncoder` for categorical data is problematic unless your data is ordinal which means that your feature has some inherent ordering. For example encoding good, ok, bad as 2, 1, 0 is OK because the scale corresponds to some ordering in the concept. However encoding John, Jane, Julie as 0, 1, 2 does not make sense since the data is nominal. This may not be an issue for modeling sometimes but it can lead to erroneous models and makes interpretation faulty. Use One Hot Encoding to avoid this problem.

I've written about this [here](https://medium.com/@amirziai/a-flask-api-for-serving-scikit-learn-models-c8bcdaa41daa#.j0j35dryb) if you're interested in learning more. Here' another [good resource](https://www.quora.com/In-scikit-learn-what-is-the-best-way-to-handle-categorical-features-of-high-cardinality-Using-one-hot-encoder-seems-to-blow-up-my-feature-space-Does-assigning-a-number-to-each-cardinality-work-better-or-do-we-run-into-the-risk-of-the-ML-algorithm-assuming-some-ordering-in-these-numeric-labels). For larger datasets OHE is computationally prohibitive and you may want to use the [hashing trick](http://scikit-learn.org/stable/modules/generated/sklearn.feature_extraction.FeatureHasher.html).

File: Projects/18_Clustering/Predict_Sales.py [hunk 3]
```diff
@@ -0,0 +1,36 @@
+# import data for offers
+import pandas as pd
+df_offers = pd.read_excel("./WineKMC.xlsx", sheetname=0)
+df_offers.columns = ["offer_id", "campaign", "varietal", "min_qty", "discount", "origin", "past_peak"]
+
+# import data for transactions
+df_transactions = pd.read_excel("./WineKMC.xlsx", sheetname=1)
+df_transactions.columns = ["customer_name", "offer_id"]
```
Comment (amirziai) [inline Projects/18_Clustering/Predict_Sales.py]: Alternatively you can pass this list to the argument `names` to `read_excel`.

File: Projects/18_Clustering/Predict_Sales.py [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+df_transactions['n'] = 1
+
+# estimate total sales per offer_id
+df_total_sales = pd.merge(df_offers['offer_id'].to_frame(), df_transactions)
```
Comment (amirziai) [inline Projects/18_Clustering/Predict_Sales.py]: If you want to extract a dataframe (even if it's one column) you can index by a list, for example  `df_offers[['offer_id]]` so you don't need to later turn it into a frame.

File: Projects/18_Clustering/Predict_Sales.py [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+df_total_sales = df_total_sales.groupby(['offer_id']).sum()
```
Comment (amirziai) [inline Projects/18_Clustering/Predict_Sales.py]: You can do your aggregation by `offer_id` directly `offer_agg = df_transactions.groupby(['offer_id']).size()` and then merge with `df_offers` like this:
`df_offers.merge(offer_agg.to_frame(name='Total Sales'), left_index=True, right_index=True)`

File: Projects/18_Clustering/Predict_Sales.py [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+df_total_sales = df_total_sales.reset_index()
+
+# create final DF
+df_offers_and_sales = pd.merge(df_offers,df_total_sales)
+
+# encode categorical predictors
+from sklearn import preprocessing
+le = preprocessing.LabelEncoder()
+
+df = df_offers_and_sales.iloc[:,1:]
```
Comment (amirziai) [inline Projects/18_Clustering/Predict_Sales.py]: For improved readability you can use this syntax to remove a list of columns:
`df[df.columns.difference(['campaign'])]`

**Review by amirziai — CHANGES_REQUESTED** [review]
@JCardenasRdz great job! your code is functionally correct except for the `LabelEncoder` usage which is technically correct but might be problematic. Added a few suggesting.

File: Projects/18_Clustering/Predict_Sales.py (hunk 2 above)
Comment (JCardenasRdz) [reply Projects/18_Clustering/Predict_Sales.py]: That is important, thank for pointing that out.

File: Projects/18_Clustering/Predict_Sales.py (hunk 1 above)
Comment (JCardenasRdz) [reply Projects/18_Clustering/Predict_Sales.py]: Got it

File: Projects/18_Clustering/Predict_Sales.py [hunk 7]
```diff
@@ -4,12 +4,11 @@
 df_offers.columns = ["offer_id", "campaign", "varietal", "min_qty", "discount", "origin", "past_peak"]
 
 # import data for transactions
-df_transactions = pd.read_excel("./WineKMC.xlsx", sheetname=1)
-df_transactions.columns = ["customer_name", "offer_id"]
+df_transactions = pd.read_excel("./WineKMC.xlsx", sheetname=1,names=["customer_name", "offer_id"])
```
Comment (amirziai) [inline Projects/18_Clustering/Predict_Sales.py]: Add a space after commas. Here's a good code [style guide](http://docs.python-guide.org/en/latest/writing/style/) for Python. Also consider using [PyCharm](https://www.jetbrains.com/pycharm/download/) as your IDE for non-Jupyter work.

**Review by amirziai — APPROVED** [review]
Looks pretty good. Left a styling comment.

**Comment by JCardenasRdz** [comment]
Thank you Amir.
